### ISSUE TITLE
Mirthsync giving false positive push in AWS CodeBuild Container

### ISSUE THREAD

**Issue by RyianRockers** [SagaHealthcareIT/mirthsync#24]
I am using a Jenkins job to pull the MirthConnect config files from a BitBucket repo, zip up the files and send them to an AWS CodeBuild Ubuntu container. I am using (and have used) the 2.0.10 version of Mirthsync. From the container I am using a push command to a MirthConnect instance in the same AWS region. The output from Mirthsync says that it has uploaded the files without any issues, but when I go to check the instance, the files that Mirthsync said it uploaded, the changes aren't there. 

![image](https://user-images.githubusercontent.com/17551097/129259760-b79735bf-50dd-459b-82fe-69720114657f.png)

The target directory $CODEBUILD_SOURCE/src directory has the file structure:
![image](https://user-images.githubusercontent.com/17551097/129260038-a85fd816-9538-45d8-8ca2-832bcfbd9010.png)

**Comment by jessesaga** [comment]
Based on your screenshot it appears that mirthSync doesn't actually find the correct files to upload. The "Uploading....." lines should be followed by an actual list of files found in the target directory location. Here's an example of what a successful push looks like.
![Screenshot from 2021-09-21 16-30-36](https://user-images.githubusercontent.com/27866363/134243349-892168fd-ca49-4e33-a4df-2b789020b059.png)

I can see how the language is confusing. I'll update the code to print a message if no files are found to push for each item (Channels, CodeTemplates, etc)

**Comment by RyianRockers** [comment]
Do you know why it wouldn't be able to find them? Could it be a permissions issue? The files are pulled from a BitBucket repo and placed on the container, so they're there.

**Comment by jessesaga** [comment]
What's the output of `tree $CODEBUILD_SRC` or `find $CODEBUILD_SRC`? Also `ls -al $CODEBUILD_SRC` might be helpful.

**Comment by RyianRockers** [comment]
I ran ls -R and all the files in the appropriate structure show up:

![image](https://user-images.githubusercontent.com/17551097/134737970-bfe1f6e9-345b-4df9-852f-15b41136e4ab.png)

**Comment by jessesaga** [comment]
Are you saying that mirthSync randomly (without changes to configuration or code) either doesn't find files or errors out with an exception?

**Comment by RyianRockers** [comment]
Yeah that seems to be the regular behavior for this implementation. FTR this is using mirthsync 2.0.10

**Comment by jessesaga** [comment]
What happens when you try to replicate the issue outside of the CodeBuild environment? Does it still randomly exhibit different results?

**Comment by RyianRockers** [comment]
mirthsync works fine when I run it from my windows box. Also note that containers running in AWS are linux, and the target mirth instances are Windows

**Comment by jessesaga** [comment]
Mirthsync works fine running from Linux to Windows and running within containers. The inconsistent behavior you're seeing reminds me of a race condition. Does some part of your buildspec.yml have async steps or are there parallel builds happening?

**Comment by RyianRockers** [comment]
Interesting, no there aren't any asynchronous steps or concurrent builds happening. Here are the contents of the buildspec file for reference:

![image](https://user-images.githubusercontent.com/17551097/134964729-47a7809a-e15b-4700-9713-7c7d905a033c.png)

**Comment by jessesaga** [comment]
The buildspec looks simple enough. Without being able to replicate this for myself.....I'm running a bit low on troubleshooting ideas. What version(s) of Mirth Connect are you pulling from and pushing to? 

You could try the latest pre-release [2.1.0](https://github.com/SagaHealthcareIT/mirthsync/releases/tag/2.1.0) and see if that makes a difference. I'd also recommend trying to start with a clean target directory (if you can) and run a mirthsync pull to repopulate it before pushing to the destination mirth server.

**Comment by RyianRockers** [comment]
I'll give that version a shot and report the results. We're trying to version control our mirth config so pulling from an existing server doesn't exactly fit the use case for deployment. We have another jenkins job that uses mirthsync to pull config and push the config to a repo and it works fine.

**Comment by jessesaga** [comment]
Ok, sounds good. Just make sure that the version of mirthsync is the same for push->git repo and the git repo -> codebuild push.

**Comment by RyianRockers** [comment]
It appears the issue is still happening:

![image](https://user-images.githubusercontent.com/17551097/134984801-3de588bb-6172-4d43-bfa0-2f4ba39c1010.png)

**Comment by jessesaga** [comment]
You could try to use the new 2.1.0 'restrict-to-path' feature to only push certain folders. This could help if the problem is in the xml or other files in the target directory. 

Another thing you could try is supplying multiple -v (-vvv) options to mirthsync to increase the log verbosity. The output will be very noisy but could reveal something.

**Comment by RyianRockers** [comment]
I haven't tried the 'restrict-to-path' feature yet, any I know this may be outside the initial problem, but this 500 cluj error does pop up from time to time. This was the result after using 2.1.0

![image](https://user-images.githubusercontent.com/17551097/135142917-8d9552d3-db1d-4cfb-881b-dfcf73beddb6.png)

**Comment by RyianRockers** [comment]
This 500 error seems to be the new behavior. I can't seem to get past it.

**Comment by jessesaga** [comment]
What version of Mirth are you using?

**Comment by RyianRockers** [comment]
2.1.0

**Comment by jessesaga** [comment]
2.1.0 is your mirthSync version. What version of Mirth Connect are you working with?

**Comment by RyianRockers** [comment]
The mirth endpoint I am hitting is 3.11. This error also occurs with 3.9

**Comment by jessesaga** [comment]
It should work fine with both 3.9 and 3.11. Is the target directory a git checkout that contains the .git directory and perhaps other extraneous or hidden files that are not the result of a 'mirthsync pull' command?

**Comment by RyianRockers** [comment]
The target directory is always from the src folder of the git repo. The src folder is the result of a pull from another environment. The structure of the src folder is standard:

<html>
<body>


/codebuild/output/src820/src/bitbucket.org/tellusolutions/tellus-mirth-connect/src:
--
117 | Channels
118 | CodeTemplates
119 | GlobalScripts
120 | configurationMap.properties


</body>
</html>

**Comment by jessesaga** [comment]
I found a bug that may be related to the behavior that you're seeing. The bug is fixed in 2.1.1 - https://github.com/SagaHealthcareIT/mirthsync/releases/tag/2.1.1

**Comment by RyianRockers** [comment]
Thank you, I'll give it a shot!

**Comment by jcdlbs** [comment]
Any update about whether 2.1.1 fixed the issue?
